src/core/lpParser.mly

```diff
@@ -0,0 +1,417 @@
+    ASSIGN de=term? SEMICOLON
+      {
+        let sym =
+          {p_sym_mod=ms; p_sym_nam=fst s; p_sym_arg=al; p_sym_typ=ao;
+           p_sym_trm=de; p_sym_prf=None; p_sym_def=true}
+        in
+        make_pos $loc (P_symbol(sym))
+      }
+  | ms=modifier* SYMBOL s=ident al=arg* ao=preceded(COLON, term)?
+    ASSIGN de=term? p=proof
+      {
+        let sym =
+          {p_sym_mod=ms; p_sym_nam=fst s; p_sym_arg=al; p_sym_typ=ao;
+           p_sym_trm=de; p_sym_prf=Some(p); p_sym_def=true}
+        in
+        make_pos $loc (P_symbol(sym))
+      }
+  | ms=modifier* INDUCTIVE is=separated_nonempty_list(WITH, inductive)
+    SEMICOLON
+      { make_pos $loc (P_inductive(ms, is)) }
+  | RULE rs=separated_nonempty_list(WITH, rule) SEMICOLON
+      { make_pos $loc (P_rules(rs)) }
+  | SET c=config SEMICOLON { make_pos $loc (P_set(c)) }
+  | q=query SEMICOLON { make_pos $loc (P_query(q)) }
+  | EOF { raise End_of_file }
+
+// Environment of a metavariable or rewrite pattern
+env: L_SQ_BRACKET ts=separated_list(SEMICOLON, term) R_SQ_BRACKET { ts }
+
+sterm:
+  // Explicit identifier (with @)
+  | AT qid=qident { make_pos $loc (P_Iden(qid, true)) }
+  // Qualified identifier
+  | qid=qident { make_pos $loc (P_Iden(qid, false)) }
+  // The wildcard "_"
+  | WILD { make_pos $loc P_Wild }
+  // The constant [TYPE] (of type Kind)
+  | TYPE { make_pos $loc P_Type }
+  // Metavariable
+  | QUESTION_MARK m=ident e=env?
+      { make_pos $loc (P_Meta(fst m, Option.map Array.of_list e)) }
+  // Pattern of rewrite rules
+  | p=patt e=env? { make_pos $loc (P_Patt(p, Option.map Array.of_list e)) }
+  // Parentheses
+  | L_PAREN t=term R_PAREN { make_pos $loc (P_Wrap(t)) }
+  // Explicitly given argument
+  | L_CU_BRACKET t=term R_CU_BRACKET { make_pos $loc (P_Expl(t)) }
+  // Natural number
+  | n=INT { make_pos $loc (P_NLit(n)) }
+
+// Applied terms
+aterm:
+  | t=aterm u=sterm { make_pos $loc (P_Appl(t,u)) }
+  | t=sterm { t }
+
+term:
+  | t=aterm { t }
+  | t=term ARROW u=term { make_pos $loc (P_Impl(t, u)) }
+  // Quantifier
+  | BACKQUOTE q=term_ident b=binder { make_pos $loc (P_Appl(q, b)) }
+  | PI xs=arg+ COMMA b=term { make_pos $loc (P_Prod(xs, b)) }
+  | PI x=arg_ident COLON a=term COMMA b=term
+      { make_pos $loc (P_Prod([[x], Some(a), false], b)) }
+  | LAMBDA xs=arg+ COMMA t=term { make_pos $loc (P_Abst(xs, t)) }
+  | LAMBDA x=arg_ident COLON a=term COMMA t=term
+      { make_pos $loc (P_Abst([[x], Some(a), false], t)) }
+  /* REVIEW: allow pattern of the form \x y z: N, t */
+  /* | LAMBDA xs=arg_ident+ COLON a=term COMMA t=term */
+  /*     { make_pos $loc (P_Abst([xs, Some(a), false], t)) } */
+  | LET x=ident a=arg* b=preceded(COLON, term)? ASSIGN t=term IN u=term
+      { make_pos $loc (P_LLet(fst x, a, b, t, u)) }
+
+binder:
+  | id=ident COMMA t=term
+      { make_pos $loc (P_Abst([[Some(fst id)], None, false], t)) }
+  | id=ident COLON a=term COMMA t=term
+      { make_pos $loc (P_Abst([[Some(fst id)], Some(a), false], t)) }
```

Comment:
I propose you to rename arg into arg_list and factorize the BACKQUOTE, PI and LAMBDA cases as follows:
```
term:
  | BACKQUOTE q=term_ident b=binder
    { make_pos $loc (P_Appl(q, Pos.make b.pos (P_Abst(fst b.elt, snd b.elt)))) }
  | PI b=binder { make_pos $loc (P_Prod(fst b.elt, snd b.elt)) }
  | LAMBDA b=binder { make_pos $loc (P_Abst(fst b.elt, snd b.elt)) }
  ...
binder:
  | xs=arg_list+ COMMA t=term { make_pos $loc (xs, t) }
  | x=arg_ident COLON a=term COMMA t=term
    { make_pos $loc ([[x], Some(a), false], t) }
```